File: civitasRUBY/lib/civitas_juego.rb
# encoding:utf-8 

require_relative 'casilla'
require_relative 'casilla_calle'
require_relative 'casilla_sorpresa'
require_relative 'casilla_impuesto'
require_relative 'casilla_juez'
require_relative 'sorpresa'
require_relative 'sorpresa_por_jugador'
require_relative 'sorpresa_por_casa_hotel'
require_relative 'sorpresa_ir_casilla'
require_relative 'sorpresa_ir_carcel'
require_relative 'sorpresa_salir_carcel'
require_relative 'sorpresa_pagar_cobrar'
require_relative 'sorpresa_jugador_especulador'
require_relative 'titulo_propiedad'
require_relative 'tablero'
require_relative 'jugador'
require_relative 'gestor_estados'
require_relative 'Operaciones_juego'

module Civitas
  class CivitasJuego
    attr_reader :indiceJugadorActual
    
    def initialize(nombres)
      @indiceJugadorActual=0;
      @jugadores=[];
      @tablero=nil
      #Rellenamos el vector de jugadores con los nombres que se pasan
      i=0;
      while i<nombres.length()
          @jugadores << Jugador.new(nombres[i]);
          i=i+1;
      end
      
      @gestorEstados=Gestor_estados.new();
      @estado=@gestorEstados.estado_inicial();
      
      @indiceJugadorActual=Dado.instance.quienEmpieza(@jugadores.size());
      
      @mazo=MazoSorpresas.new();
      inicializaTablero(@mazo);
      inicializaMazoSorpresas(@tablero);
    end
    
    def actualizarInfo()
      puts ((@jugadores[@indiceJugadorActual]).toString)
      
      if(finalDelJuego())
        nombre_jugador_actual=getJugadorActual().nombre
        rank=[]
        rank=ranking()
            
        puts "\n ***Se ha acabado el juego, el jugador #{nombre_jugador_actual} esta en bancarrota***"
      
        tamanio=rank.size()
        case (tamanio)
              when 4
                    puts "\n\n *Ranking*
                 Primer puesto: #{(rank[tamanio-1]).nombre} con saldo: #{(rank[tamanio-1]).saldo}
                 Segundo puesto: #{(rank[tamanio-2]).nombre} con saldo: #{(rank[tamanio-2]).saldo}
                 Tercer puesto: #{(rank[tamanio-3]).nombre} con saldo: #{(rank[tamanio-3]).saldo}
                 Cuarto puesto: #{(rank[tamanio-4]).nombre} con saldo: #{(rank[tamanio-4]).saldo}"
                   
              when 3
                    puts "\n\n *Ranking*
                 Primer puesto: #{(rank[tamanio-1]).nombre} con saldo: #{(rank[tamanio-1]).saldo}
                 Segundo puesto: #{(rank[tamanio-2]).nombre} con saldo: #{(rank[tamanio-2]).saldo}
                 Tercer puesto: #{(rank[tamanio-3]).nombre} con saldo: #{(rank[tamanio-3]).saldo}"
                    
              when 2
                    puts "\n\n *Ranking*
                 Primer puesto: #{(rank[tamanio-1]).nombre} con saldo: #{(rank[tamanio-1]).saldo}
                 Segundo puesto: #{(rank[tamanio-2]).nombre} con saldo: #{(rank[tamanio-2]).saldo}"
                    
      end
      end
    end
    
    def salirCarcelPagando()
      return @jugadores[@indiceJugadorActual].salirCarcelPagando;
    end
    
    def salirCarcelTirando()
      return @jugadores[@indiceJugadorActual].salirCarcelTirando;
    end
    
    def siguientePaso()
      jugadorActual=@jugadores[@indiceJugadorActual]
      operacion=@gestorEstados.operaciones_permitidas(jugadorActual, @estado)
      puts operacion.to_s
      case(operacion)
        when Operaciones_juego::PASAR_TURNO
          pasarTurno()
          siguientePasoCompletado(operacion)
          
        when Operaciones_juego::AVANZAR
          avanzaJugador()
          siguientePasoCompletado(operacion)
      end
      return operacion
    end
    
    def siguientePasoCompletado(operacion)
      @estado=@gestorEstados.siguiente_estado(@jugadores[@indiceJugadorActual], @estado, operacion)
    end
    
    def construirCasa(ip)
      return (@jugadores[@indiceJugadorActual]).construirCasa(ip)
    end
    
    def construirHotel(ip)
      return (@jugadores[@indiceJugadorActual]).construirHotel(ip)
    end
    
    def vender(ip)
      return (@jugadores[@indiceJugadorActual].vender(ip))
    end
    
    def comprar()
      jugadorActual=@jugadores[@indiceJugadorActual]
      numCasillaActual=jugadorActual.numCasillaActual
      casilla=@tablero.getCasilla(numCasillaActual)
      titulo=casilla.titulo
      res=jugadorActual.comprar(titulo)
      
      return res
    end
    
    def hipotecar(ip)
      return (@jugadores[@indiceJugadorActual]).hipotecar(ip)
    end
    
    def cancelarHipoteca(ip)
      return (@jugadores[@indiceJugadorActual]).cancelarHipoteca(ip)
    end
    
    def finalDelJuego()
      haybancarrota_ya=false
      i=0
      while(i<@jugadores.size and !haybancarrota_ya)
        if(@jugadores[i].enBancarrota)
          haybancarrota_ya=true
        end
        i=i+1
      end
      return haybancarrota_ya
    end
    
    #En el guion se llama getCasillaActual
    def getCasillaActual()
      return @tablero.getCasilla(@jugadores[@indiceJugadorActual].numCasillaActual)
    end
    
    def getJugadorActual()
      return @jugadores[@indiceJugadorActual]
    end

    
    private    
    def contabilizarPasosPorSalida(jugadorActual)
      while(@tablero.getPorSalida()>0)        
            jugadorActual.pasaPorSalida();
      end
    end
    
    def pasarTurno()
      if(@indiceJugadorActual>=0 and @indiceJugadorActual<@jugadores.size()-1)
        @indiceJugadorActual=@indiceJugadorActual+1
      else
        @indiceJugadorActual=0
      end
    end
    
    
    def inicializaTablero(mazo)
        casillaCarcel = 4
        @tablero = Tablero.new(casillaCarcel) #En el initialize de tablero se aniade la casilla SALIDA en la posicion 0                                                                # Salida. Posicion 0
        
        @tablero.aniadeCasilla(CasillaCalle.constructorCasillaCalle("BAR DE LUCHO", TituloPropiedad.new("BAR DE LUCHO", 40, 1.1, 80, 120, 30), 120))                                   #Calle 1. Posicion 1
        @tablero.aniadeCasilla(CasillaCalle.constructorCasillaCalle("CHUBBY-ESCUELA", TituloPropiedad.new("CHUBBY-ESCUELA", 45, 1.1, 90, 135, 30), 135))                               #Calle 2. Posicion 2
        @tablero.aniadeCasilla(CasillaSorpresa.constructorCasillaSorpresa("SORPRESA", mazo))                                                                                           #SORPRESA. Posicion 3
       #En la llamada anterior como la carcel es el 4 en el vector, el metodo lo comprueba y añade, no hay que hacerlo explicitamente.                                                 # Carcel. Posicion 4
        @tablero.aniadeCasilla(CasillaCalle.constructorCasillaCalle("CALLE DE LA LOLA",TituloPropiedad.new("CALLE DE LA LOLA", 50, 1.2, 100, 150, 40),150))                            #Calle 3. Posicion 5
        @tablero.aniadeCasilla(CasillaCalle.constructorCasillaCalle("CALLE KALASHNIKOV",TituloPropiedad.new("CALLE KALASHNIKOV", 55, 1.2, 110, 160, 40),160))                          #Calle 4. Posicion 6
        @tablero.aniadeCasilla(CasillaCalle.constructorCasillaCalle("CALLE DE LAS FICHAS",TituloPropiedad.new("CALLE DE LAS FICHAS", 60, 1.2, 120, 175, 45),175))                      #Calle 5. Posicion 7
        @tablero.aniadeCasilla(CasillaImpuesto.constructorCasillaImpuesto("IMPUESTO", 500))                                                                                            #IMPUESTO. Posicion 8
        @tablero.aniadeCasilla(CasillaCalle.constructorCasillaCalle("CALLE DE LOS PATOS",TituloPropiedad.new("CALLE DE LOS PATOS", 65, 1.3, 130, 180, 50),180))                        #Calle 6. Posicion 9
        @tablero.aniadeJuez()                                                                                                                                                          #JUEZ. Posicion 10
        @tablero.aniadeCasilla(CasillaSorpresa.constructorCasillaSorpresa("SORPRESA", mazo))                                                                                           #SORPRESA. Posicion 11
        @tablero.aniadeCasilla(CasillaCalle.constructorCasillaCalle("CALLE DEL RGB",TituloPropiedad.new("CALLE DEL RGB", 70, 1.3, 140, 200, 60),200))                                  #Calle 7. Posicion 12
        @tablero.aniadeCasilla(CasillaCalle.constructorCasillaCalle("AVENIDA DE ESPANHA",TituloPropiedad.new("AVENIDA DE ESPANHA", 80, 1.4, 150, 220, 75),220))                        #Calle 8. Posicion 13
        @tablero.aniadeCasilla(CasillaCalle.constructorCasillaCalle("CALLE DEL PINCHAZO",TituloPropiedad.new("CALLE DEL PINCHAZO", 90, 1.4, 160, 240, 85),240))                        #Calle 9. Posicion 14
        @tablero.aniadeCasilla(Casilla.constructorCasilla("PARKING"))                                                                                                                  #PARKING. Posicion 15
        @tablero.aniadeCasilla(CasillaCalle.constructorCasillaCalle("CALLE DE LEO",TituloPropiedad.new("CALLE DE LEO", 100, 1.4, 170, 250, 85),250))                                   #Calle 10. Posicion 16
        @tablero.aniadeCasilla(CasillaSorpresa.constructorCasillaSorpresa("SORPRESA", mazo))                                                                                           #SORPRESA. Posicion 17
        @tablero.aniadeCasilla(CasillaCalle.constructorCasillaCalle("FACULTAD DE BELLAS ARTES",TituloPropiedad.new("FACULTAD DE BELLAS ARTES", 120, 1.5, 200, 275, 100),275))          #Calle 11. Posicion 18
        @tablero.aniadeCasilla(CasillaCalle.constructorCasillaCalle("TERCER EDIFICIO DE LA ETSIIT",TituloPropiedad.new("TERCER EDIFICIO DE LA ETSIIT", 25, 1.5, 200, 300, 100),300))   #Calle 12. Posicion 19
    end
    
    def inicializaMazoSorpresas(unTablero)

        # Las dos casillas PAGARCOBRAR
        @mazo.alMazo(SorpresaPagarCobrar.constructorSorpresaPagarCobrar("Te han dado una subvención por tontico, recibes 500", 500))
        @mazo.alMazo(SorpresaPagarCobrar.constructorSorpresaPagarCobrar("Te robado el iPhone, pierdes 500",-500))
        
        # Las dos casillas IRACASILLA
        @mazo.alMazo(SorpresaIrCasilla.constructorSorpresaIrCasilla("Vas a construir el tercer edificio de la ETSIIT, ve a la casilla 19",unTablero,19))
        @mazo.alMazo(SorpresaIrCasilla.constructorSorpresaIrCasilla("Te han tirado full de fichas, sientente afortunado y muevete a la casilla 7",unTablero,7))

        # Para ir a la carecl y salir de ella
        @mazo.alMazo(SorpresaIrCarcel.constructorSorpresaIrCarcel("Te ha pillado el FBI, vas para la carcel",unTablero))
        @mazo.alMazo(SorpresaSalirCarcel.constructorSorpresaSalirCarcel("Conseguiste un pagare, tendras un favor con los cops",@mazo))

        # Pagar o cobrar por cada casa y hotel
        @mazo.alMazo(SorpresaPorCasaHotel.constructorSorpresaPorCasaHotel("Eres un fiera inmobiliario, cobras 200 por casa/hotel",200))
        @mazo.alMazo(SorpresaPorCasaHotel.constructorSorpresaPorCasaHotel("El seguro te ha estafado, pierdes 200 por casa/hotel",-200))
        
        # Para cobrar o pagar de/a cada jugador
        @mazo.alMazo(SorpresaPorJugador.constructorSorpresaPorJugador("Compartir es vivir, cada jugador te regala 100",100))
        @mazo.alMazo(SorpresaPorJugador.constructorSorpresaPorJugador("Perdiste una apuesta con tus amigos imaginarios, pierdes 100 por cada uno de ellos",-100))

        # Para cambiar a Jugador Especulador
        @mazo.alMazo(SorpresaJugadorEspeculador.constructorSorpresaJugadorEspeculador("Congratulaciones, se ha convertido usted en un especulador",200))
        
    end
    
    def ranking()
      @jugadores.sort! #Me ordena el array en el propio array(osea me lo modifica), usando el operador <=>
      return @jugadores
    end
    
    def avanzaJugador()
      jugadorActual=@jugadores[@indiceJugadorActual]
      posicionActual=jugadorActual.numCasillaActual
      tirada=Dado.instance.tirar
      
      #Esto no lo dice el guion, pero me parece util saber la tirada que ha salido
      Diario.instance.ocurre_evento("\tAl jugador #{getJugadorActual.nombre} le ha salido un #{tirada} en el dado")
      
      posicionNueva=@tablero.nuevaPosicion(posicionActual, tirada)
      casilla=@tablero.getCasilla(posicionNueva)
      contabilizarPasosPorSalida(jugadorActual)
      jugadorActual.moverACasilla(posicionNueva)
      casilla.recibeJugador(@indiceJugadorActual,@jugadores)
      contabilizarPasosPorSalida(jugadorActual)
    end
    
  end
end

File: civitasJAVA/src/civitas/CivitasJuego.java
package civitas;

import java.util.ArrayList;
import GUI.*;

/**
 * @author Jesús Navarro Merino
 * @author José Alberto Gómez García
 */

public class CivitasJuego {
    private int indiceJugadorActual;
    private Tablero tablero;
    private MazoSorpresas mazo;
    private ArrayList<Jugador> jugadores;
    private EstadosJuego estado;
    private GestorEstados gestorEstados;
    
    
    
    public CivitasJuego(ArrayList<String> nombres){
        jugadores= new ArrayList<>();
        for(int i=0; i<nombres.size(); i++)
            jugadores.add(new Jugador(nombres.get(i)));
        
        gestorEstados= new GestorEstados();
        estado=gestorEstados.estadoInicial();
        
        indiceJugadorActual=Dado.getInstance().quienEmpieza(jugadores.size());
        
        //Crear el mazo de sorpresas
        //Llamar al método de inicialización del tablero.
        //Llamar al método de inicialización del mazo de sorpresas.
        mazo= new MazoSorpresas();
        inicializaTablero(mazo);
        inicializaMazoSorpresas(tablero);
    }
    
    public String stringRanking(){
        ArrayList<Jugador> rank;
        //Esta variable local "nombre_jugador_actual" la creo porque dentro del metodo ranking() se llama a sort(), que me ordena
        //el ArrayList "jugadores" (sobre él mismo, osea que me lo modifica). Por tanto, si luego quisiera hacer un
        // "getJugadorActual().nombre" no me cogería el jugador que deberia ser
        String nombre_jugador_actual=getJugadorActual().nombre;
        rank=ranking();
            
        switch (rank.size()){
            case 4:
                return("\n ***Se ha acabado el juego, el jugador " + nombre_jugador_actual + " esta en bancarrota***"
                    + "\n\n *Ranking*"
                        + "\n\t Primer puesto: " + rank.get(rank.size()-1).nombre + " con saldo: " + rank.get(rank.size()-1).getSaldo()
                        + "\n\t Segundo puesto: " + rank.get(rank.size()-2).nombre + " con saldo: " + rank.get(rank.size()-2).getSaldo()
                        + "\n\t Tercer puesto: " + rank.get(rank.size()-3).nombre + " con saldo: " + rank.get(rank.size()-3).getSaldo()
                        + "\n\t Cuarto puesto: " + rank.get(rank.size()-4).nombre + " con saldo: " + rank.get(rank.size()-4).getSaldo());
     
            case 3:
                return("\n ***Se ha acabado el juego, el jugador " + nombre_jugador_actual + " esta en bancarrota***"
                    + "\n\n *Ranking*"
                        + "\n\t Primer puesto: " + rank.get(rank.size()-1).nombre + " con saldo: " + rank.get(rank.size()-1).getSaldo()
                        + "\n\t Segundo puesto: " + rank.get(rank.size()-2).nombre + " con saldo: " + rank.get(rank.size()-2).getSaldo()
                        + "\n\t Tercer puesto: " + rank.get(rank.size()-3).nombre + " con saldo: " + rank.get(rank.size()-3).getSaldo());
             
            case 2:
                return("\n ***Se ha acabado el juego, el jugador " + nombre_jugador_actual + " esta en bancarrota***"
                    + "\n\n *Ranking*"
                        + "\n\t Primer puesto: " + rank.get(rank.size()-1).nombre + " con saldo: " + rank.get(rank.size()-1).getSaldo()
                        + "\n\t Segundo puesto: " + rank.get(rank.size()-2).nombre + " con saldo: " + rank.get(rank.size()-2).getSaldo());
        }
        //Esto me lo dice el IDE que lo ponga sino da error
        return null;
    }
    
    
    private void inicializaTablero(MazoSorpresas mazo){
            int casillaCarcel = 5;
        tablero = new Tablero(casillaCarcel);   //Creamos la casilla de la salida (posicion 0) e indicamos el numero de la carcel.                                    //Salida.   Posicion 0
        
        tablero.añadeCasilla(new CasillaCalle("Bar de Lucho",new TituloPropiedad("Bar de Lucho", 80, (float) 1.1, 160, 240, 60),240));                                //Calle 1.  Posicion 1
        tablero.añadeCasilla(new CasillaCalle("Chubby-Escuela",new TituloPropiedad("Chubby-Escuela", 90, (float) 1.1, 180, 270, 60),270));                            //Calle 2.  Posicion 2
        tablero.añadeCasilla(new CasillaSorpresa(mazo, "Sorpresa1"));                                                                                                 //Sorpresa1.Posicion 3                                                                                                                         //Carcel.   Posicion 4
        tablero.añadeCasilla(new CasillaCalle("Ayuntamiento de Jodar",new TituloPropiedad("Ayuntamiento de Jodar", 100, (float) 1.2, 200, 300, 80),300));                             //Calle 3.  Posicion 4
        //Al aniadir el Ayto de Jodar, como la carcel tiene la posicion 5, el metodo lo comprueba y aniade,                                                           //Carcel.   Posicion 5
        //no hay que hacerlo explicitamente.
        tablero.añadeCasilla(new CasillaCalle("Calle Ubuntu",new TituloPropiedad("Calle Ubuntu", 110, (float) 1.2, 220, 320, 80),320));                               //Calle 4.  Posicion 6
        tablero.añadeCasilla(new CasillaCalle("Calle de las Fichas",new TituloPropiedad("Calle de las Fichas", 120, (float) 1.2, 240, 350, 90),350));                 //Calle 5.  Posicion 7
        tablero.añadeCasilla(new CasillaImpuesto("Impuesto",(float) 500.0));                                                                                          //Impuesto. Posicion 8
        tablero.añadeCasilla(new CasillaCalle("Calle de los Patos",new TituloPropiedad("Calle de los Patos", 130, (float) 1.3, 260, 360, 100),360));                  //Calle 6.  Posicion 9
        tablero.añadeCasilla(new Casilla("Parking"));                                                                                                                 //Parking.  Posicion 10
        tablero.añadeCasilla(new CasillaCalle("Calle del RGB",new TituloPropiedad("Calle del RGB", 140, (float) 1.3, 280, 400, 120),400));                            //Calle 7.  Posicion 12
        tablero.añadeCasilla(new CasillaSorpresa(mazo,"Sorpresa2"));                                                                                                  //Sorpresa2.Posicion 11
        tablero.añadeCasilla(new CasillaCalle("Avenida de España",new TituloPropiedad("Avenida de España", 160, (float) 1.4, 300, 440, 150),440));                    //Calle 8.  Posicion 13
        tablero.añadeCasilla(new CasillaCalle("Calle del Pinchazo",new TituloPropiedad("Calle del Pinchazo", 180, (float) 1.4, 320, 480, 170),480));                  //Calle 9.  Posicion 14
        tablero.añadeJuez();                                                                                                                                          //Juez.     Posicion 15
        tablero.añadeCasilla(new CasillaCalle("Calle de Leo",new TituloPropiedad("Calle de Leo", 200, (float) 1.4, 240, 500, 170),500));                              //Calle 10. Posicion 16
        tablero.añadeCasilla(new CasillaSorpresa(mazo, "Sorpresa3"));                                                                                                 //Sorpresa3.Posicion 17
        tablero.añadeCasilla(new CasillaCalle("Bellas Artes",new TituloPropiedad("Bellas Artes", 240, (float) 1.5, 400, 550, 200),550));                              //Calle 11. Posicion 18
        tablero.añadeCasilla(new CasillaCalle("3º edificio Etsiit",new TituloPropiedad("3º edificio Etsiit", 260, (float) 1.5, 400, 650, 250),650));                        //Calle 12. Posicion 19

        
    }
    
    //Se crea el mazo de sorpresas
    //LO DE ABAJO SON LAS REGLAS DEL JUEGO (MIRAR TABLA)
    //Cuando un jugador cae en una casilla de tipo SORPRESA debe coger una carta sorpresa y hacer lo
    //que se indique en ella. En el mazo de cartas Sorpresa hay 10 cartas de 5 tipos diferentes como
    //muestra la tabla siguiente. En la mayoría de ellas se necesita conocer un valor indicado en la propia
    //carta cuya interpretación es diferente según el tipo de carta.
    
    
    private void inicializaMazoSorpresas(Tablero tablero){ 

         //Las dos casillas PAGARCOBRAR
         mazo.alMazo(new SorpresaPagarCobrar("Te han dado una subvención por tontico, recibes 500",500));
         mazo.alMazo(new SorpresaPagarCobrar("Te robado el iPhone, pierdes 500",-500));
         
         //Las dos casillas IRACASILLA
         mazo.alMazo(new SorpresaIrCasilla("Vas a construir el 3º edificio de la ETSIIT, ve a la casilla 19",tablero,19));
         mazo.alMazo(new SorpresaIrCasilla("Te han tirado full de fichas, sientente afortunado y muevete a la casilla 7",tablero,7));
         
         //Para ir a la carcel y salir de ella
         mazo.alMazo(new SorpresaIrCarcel("Te ha pillado el FBI, vas para la carcel",tablero));
         mazo.alMazo(new SorpresaSalirCarcel("Conseguiste un pagare, tendras un favor con los cops",mazo)); 
                
         //Pagar o cobrar por cada casa y hotel
         mazo.alMazo(new SorpresaPorCasaHotel("Eres un fiera inmobiliario, cobras 200 por casa/hotel",200));
         mazo.alMazo(new SorpresaPorCasaHotel("El seguro te ha estafado, pierdes 200 por casa/hotel",-200));
 
         //Para cobrar o pagar a cada jugador
         mazo.alMazo(new SorpresaPorJugador("Compartir es vivir, cada jugador te regala 100",100));
         mazo.alMazo(new SorpresaPorJugador("Perdiste una apuesta con tus amigos imaginarios, pierdes 100 por cada uno de ellos",-100));   
         
         //Para cambiar a JugadorEspeculador
         mazo.alMazo(new SorpresaJugadorEspeculador("Congratulaciones, se ha convertido usted en un especulador",200));
    }
    
    private void contabilizarPasosPorSalida (Jugador jugadorActual){
        while(tablero.getPorSalida()>0){ 
            jugadorActual.pasaPorSalida();
        }
    }
    
    private void pasarTurno(){
        if(indiceJugadorActual >=0 && indiceJugadorActual<jugadores.size()-1)
            indiceJugadorActual++;
        else
            indiceJugadorActual=0;
    }
    
    //El parametro solo se lo paso para poder usar el quitarBordes(), porque sino no hay otra manera o al menos yo no sé
    public OperacionesJuego siguientePaso(CivitasView vista){
        Jugador jugadorActual=jugadores.get(indiceJugadorActual);
        OperacionesJuego operacion=gestorEstados.operacionesPermitidas(jugadorActual, estado);
        
        //Tanto al pasar turno como al avanzar de casilla, quitamos el tablero que corresponda (es decir, a la 
        //casilla en la que esté el jugador actual)
        switch(operacion){
            case PASAR_TURNO:
                vista.quitarBordes();
                pasarTurno();
                siguientePasoCompletado(operacion);
                break;
                
            case AVANZAR:
                //El quitarBordes() se lo meto dentro al avanzaJugador para que quite el tablero en el momento justo (y que parezca mas interactivo)
                avanzaJugador(vista);
                siguientePasoCompletado(operacion);
                break;
        }
        return operacion;
    }
    
    public void siguientePasoCompletado (OperacionesJuego operacion){
        estado=gestorEstados.siguienteEstado(jugadores.get(indiceJugadorActual), estado, operacion);
    }
    
    public boolean construirCasa (int ip){
        return (jugadores.get(indiceJugadorActual)).construirCasa(ip);
    }
    
    public boolean construirHotel(int ip){
        return (jugadores.get(indiceJugadorActual)).construirHotel(ip);
    }
    
    public boolean vender(int ip){
         return (jugadores.get(indiceJugadorActual)).vender(ip);
    }
    
    public boolean comprar(){
        Jugador jugadorActual=jugadores.get(indiceJugadorActual);
        int numCasillaActual=jugadorActual.getNumCasillaActual();
        Casilla casilla=tablero.getCasilla(numCasillaActual);
        TituloPropiedad titulo=casilla.getTituloPropiedad();
        boolean res=jugadorActual.comprar(titulo);
        
        return res;
    }
    
    public boolean hipotecar(int ip){
        return (jugadores.get(indiceJugadorActual)).hipotecar(ip);
    }
    
    public boolean cancelarHipoteca(int ip){
        return (jugadores.get(indiceJugadorActual)).cancelarHipoteca(ip);
    }
    
    public boolean salirCarcelPagando(){
        return (jugadores.get(indiceJugadorActual)).salirCarcelPagando();
    }
    
    public boolean salirCarcelTirando(){
        return (jugadores.get(indiceJugadorActual)).salirCarcelTirando();
    }
    
    public boolean finalDelJuego(){
        boolean haybancarrota_ya=false;
        for(int i=0; i<jugadores.size() && !haybancarrota_ya; i++)
            if(jugadores.get(i).enBancarrota())
                haybancarrota_ya=true;
        return haybancarrota_ya;
    }
    
     private void avanzaJugador(CivitasView vista){
        Jugador jugadorActual=jugadores.get(indiceJugadorActual);       
        int posicionActual=jugadorActual.getNumCasillaActual();         
        int tirada=Dado.getInstance().tirar();                         
        
        //Esto no lo dice el guion, pero me parece util saber la tirada que ha salido
        Diario.getInstance().ocurreEvento("\t Al jugador " + getJugadorActual().nombre + " le ha salido un " + tirada + " en el dado");
        
        int posicionNueva=tablero.nuevaPosicion(posicionActual, tirada);
        Casilla casilla=tablero.getCasilla(posicionNueva);                
        this.contabilizarPasosPorSalida(jugadorActual);
        
        //Lo de vista.quitarBordes() lo pongo aquí para que mientras está popeada la ventanita del dado (y portanto todavia no se haya clikado el boton, 
        //y portanto todavia no se haya movido al jugador a otra casilla distinta), se pueda ver el borde en la casilla en la que esté
        //el jugador actual
        
        //Como la siguiente linea, es la de moverACasilla(posicionNueva), le quito los bordes justo antes
        vista.quitarBordes();
        
        jugadorActual.moverACasilla(posicionNueva);
        
        //Esto va bien porq: el compilador ve que "casilla" es de la Clase Casilla y tiene el metodo recibeJugador (que en el caso de las casillas Descanso solo llama al informe)
        //Lo que pasa esq en tiempo de ejecucion, el interprete ve que le estamos metiendo una casillaCalle, casillaJuez, casillaImpuesto, o casillaSorpresa (en caso de que le metamos alguna de estas, claro)
        //y llamará al metodo recibe de cada uno
        casilla.recibeJugador(indiceJugadorActual, jugadores);   
        this.contabilizarPasosPorSalida(jugadorActual);                   
     }
     
    private ArrayList<Jugador> ranking(){
        jugadores.sort(null); //Me ordena el array usando el orden especificado. Si meto null como parametro, 
        //cogerá el orden de la interface Comparable<>, y como hemos overrideado el compareTo, ordenará por saldos
	
        return jugadores; //return clasificacion
    }
    
    public Jugador getJugadorActual(){
        return jugadores.get(indiceJugadorActual);
    }
    
    
    public Casilla getCasillaActual(){
        return tablero.getCasilla((jugadores.get(indiceJugadorActual)).getNumCasillaActual());
    }
    
    public Tablero getTablero(){
        return tablero;
    }
    
    public ArrayList<Jugador> getJugadores(){
        return jugadores;
    }

}//Fin de clase



